docs:  how to create a custom Attribute decorator in a child component and use it from a parent component (#38721)
angular/angular: aio/content/guide/attribute-binding.md

Imagine that the `dirWithHostBinding` directive and the `comp-with-host-binding` component both have a `[style.width]` host binding.
In that case, if `dirWithHostBinding` sets its binding to `undefined`, the `width` property will fall back to the value of the `comp-with-host-binding` host binding.
However, if `dirWithHostBinding` sets its binding to `null`, the `width` property will be removed entirely.

> ### Create custom attributes

Author: Thanks for this!

> It is possible to create a custom attribute on a child component so that a parent can provide it with a value. The [Attribute](api/core/Attribute) decorator is defined in the component's `constructor` and accepts one argument, which is the exported name. 

Author: Yes, you are right. By this description, It's not that clear what the Attribute does. I will change completely the description to make it more straight forward

> <code-example path="attribute-binding/src/app/attribute-decorator-child.component.ts" header="attribute-decorator-child.component.ts"></code-example>

Author: got it

<code-example path="attribute-binding/src/app/app.component.html" region="attribute-decorator" header="src/app/app.component.html"></code-example>

> The result of the `app.component.html` will be **Hello world**

Author: I love the consistency comments!!! 💟

Imagine that the `dirWithHostBinding` directive and the `comp-with-host-binding` component both have a `[style.width]` host binding.
In that case, if `dirWithHostBinding` sets its binding to `undefined`, the `width` property will fall back to the value of the `comp-with-host-binding` host binding.
However, if `dirWithHostBinding` sets its binding to `null`, the `width` property will be removed entirely.

> ### Creating custom attributes

Author: Yup. I agree

> There are cases that you would need to differentiate the result of a [Component](api/core/Component) or a [Directive](api/core/Directive) based on a given value. The Host element can inject this value in the respective component/directive. Some well known cases are the HTML Button and the Input. The Button has the *type* attribute, which accepts the value "button", "reset" and "submit". A similar situation exists with the Input.
> 
> The [@Input()](api/core/Input) decorator can inject this value as a string and apply this differentiation on the component/directive. However, the value won't be accessible before the change detection runs, and, as such, you won't be able to use the value in the constructor.
> 
> If you need the Host element to inject a plain value, the [Attribute](api/core/Attribute) decorator is great for this. Please note that the injected value cannot be set or changed dynamically. 
> 
> The [RouterOutlet](api/router/RouterOutlet) makes use of the [Attribute](api/core/Attribute) decorator to provide a unique [name](api/router/RouterOutlet#description) on each outlet.
> 
> <code-example path="attribute-binding/src/app/attribute-decorator-my-input.component.ts" header="src/app/attribute-decorator-my-input.component.ts"></code-example>
> 
> <code-example path="attribute-binding/src/app/app.component.html" region="attribute-decorator" header="src/app/app.component.html"></code-example>
> 
> The result of the `app.component.html` will be **The type of the input is: number**.

Reviewer: Let me suggest an alternative layout and phrasing. (I shuffled the content around a bit, re-phrased some parts clarity, introduced idiomatic angular.io presentational elements ([alerts/callouts][1])). WDYT?  

```suggestion
There are cases where you need to differentiate the behavior of a [Component](api/core/Component) or [Directive](api/core/Directive) based on a static value set on the host element as an HTML attribute. For example, you might have a directive that needs to know the `type` of a `<button>` or `<input>` element.

The [Attribute](api/core/Attribute) parameter decorator is great for passing the value of an HTML attribute to a component/directive constructor via [dependency injection](guide/dependency-injection).

<div class="alert is-helpful">

  Note that the injected value captures the value of the specified HTML attribute at that moment.
  Future updates to the attribute value are not reflected on the injected value.

</div>

<code-example
    path="attribute-binding/src/app/attribute-decorator-my-input.component.ts"
    header="src/app/attribute-decorator-my-input.component.ts">
</code-example>

<code-example
    path="attribute-binding/src/app/app.component.html"
    region="attribute-decorator"
    header="src/app/app.component.html">
</code-example>

In the above example, the result of `app.component.html` will be **The type of the input is: number**.

Another example, is the [RouterOutlet](api/router/RouterOutlet) directive, which makes use of the [Attribute](api/core/Attribute) decorator to retrieve the unique [name](api/router/RouterOutlet#description) on each outlet.

<div class="callout is-helpful">
  <header>`@Attribute()` vs `@Input()`</header>

  The [@Input()](api/core/Input) property decorator can also be used to retrieve the value of an HTML attribute from a component's/directive's host element, but it has some differences from `@Attribute()`:
  - `@Input()` keeps track of the attribute value and updates the bound component/directive property.
  - With `@Input()`, the value won't be accessible before the first [change detection](guide/glossary#change-detection) run.

  Use the decorator that is better suited to your particular usecase.

</div>
```

[1]: https://angular.io/guide/docs-style-guide#alerts-and-callouts

Author: Well, I think that your suggestions look more the Angular way :)
Thanks also for letting me know of the presentational elements.

Imagine that the `dirWithHostBinding` directive and the `comp-with-host-binding` component both have a `[style.width]` host binding.
In that case, if `dirWithHostBinding` sets its binding to `undefined`, the `width` property will fall back to the value of the `comp-with-host-binding` host binding.
However, if `dirWithHostBinding` sets its binding to `null`, the `width` property will be removed entirely.

### Injecting attribute values

There are cases where you need to differentiate the behavior of a [Component](api/core/Component) or [Directive](api/core/Directive) based on a static value set on the host element as an HTML attribute. For example, you might have a directive that needs to know the `type` of a `<button>` or `<input>` element.

The [Attribute](api/core/Attribute) parameter decorator is great for passing the value of an HTML attribute to a component/directive constructor via [dependency injection](guide/dependency-injection).

<div class="alert is-helpful">

>   The injected value captures the value of the specified HTML attribute at moment.

Reviewer: ```suggestion
  The injected value captures the value of the specified HTML attribute at that moment.
```

Imagine that the `dirWithHostBinding` directive and the `comp-with-host-binding` component both have a `[style.width]` host binding.
In that case, if `dirWithHostBinding` sets its binding to `undefined`, the `width` property will fall back to the value of the `comp-with-host-binding` host binding.
However, if `dirWithHostBinding` sets its binding to `null`, the `width` property will be removed entirely.

### Injecting attribute values

There are cases where you need to differentiate the behavior of a [Component](api/core/Component) or [Directive](api/core/Directive) based on a static value set on the host element as an HTML attribute. For example, you might have a directive that needs to know the `type` of a `<button>` or `<input>` element.

The [Attribute](api/core/Attribute) parameter decorator is great for passing the value of an HTML attribute to a component/directive constructor via [dependency injection](guide/dependency-injection).

> <div class="alert is-helpful">
>   Note that the injected value captures the value of the specified HTML attribute at that moment.
>   Future updates to the attribute value are not reflected on the injected value.
> </div>

Reviewer 2: ```suggestion
<div class="alert is-helpful">
  The injected value captures the value of the specified HTML attribute at moment.
  Future updates to the attribute value are not reflected in the injected value.
</div>
```

<code-example
    path="attribute-binding/src/app/my-input-with-attribute-decorator.component"
    header="src/app/my-input-with-attribute-decorator.component">
</code-example>

<code-example
    path="attribute-binding/src/app/app.component.html"
    region="attribute-decorator"
    header="src/app/app.component.html">
</code-example>

> In the above example, the result of `app.component.html` will be **The type of the input is: number**.

Reviewer 2: ```suggestion
In the preceding example, the result of `app.component.html` is **The type of the input is: number**.
```

> Another example, is the [RouterOutlet](api/router/RouterOutlet) directive, which makes use of the [Attribute](api/core/Attribute) decorator to retrieve the unique [name](api/router/RouterOutlet#description) on each outlet.

Reviewer 2: ```suggestion
Another example is the [RouterOutlet](api/router/RouterOutlet) directive, which makes use of the [Attribute](api/core/Attribute) decorator to retrieve the unique [name](api/router/RouterOutlet#description) on each outlet.
```

<div class="callout is-helpful">

  <header>@Attribute() vs @Input()</header>

>   The [@Input()](api/core/Input) property decorator can also be used to retrieve the value of an HTML attribute from a component's/directive's host element, but it has some differences from `@Attribute()`:

Reviewer 2: ```suggestion
You can also use the [@Input()](api/core/Input) property decorator to retrieve the value of an HTML attribute from a component's/directive's host element. However, `@Input()` does have some differences from `@Attribute()`:
```

  - `@Input()` keeps track of the attribute value and updates the bound component/directive property.

>   - With `@Input()`, the value won't be accessible before the first [change detection](guide/glossary#change-detection) run.

Reviewer 2: ```suggestion
  - With `@Input()`, the value is not accessible before the first [change detection](guide/glossary#change-detection) run.
```

>   Use the decorator that is better suited to your particular use-case.

Reviewer 2: I recommend deleting this. How does the user know which decorator is better suited to their use case? If we can't provide guidance, we should just leave this sentence out.

Reviewer 3: But we should be able. It's one of the important things why this topic is introduced - to explain why `@Attribute` exists at all and what are the specifics when used.

Reviewer 2: I understand what you mean. But my point is: If we can provide that guidance, let's provide it! 

As an example, imagine if I wrote: "Reactive Forms behave *this* way. Template-driven forms behave *that* way. Choose the one that's right for you!" How do you know which one is right for you? It's better if we say "Reactive Forms behave *this* way. It's recommended you use reactive forms in situations such as..."

Do we have similar content regarding when it's recommended to use `@Input()` over `@Attribute()`? It may be in this topic, and I'm missing it because the preview hasn't caught up yet. :)

Reviewer 3: Actually these two points, already mentioned, are the known differences and also the guidelines why / where to use the first or the other:

\- \`@Input()\` keeps track of the attribute value and updates the bound component/directive property

\- With \`@Input()\`, the value won't be accessible before the first \[change detection\](guide/glossary#change-detection) run.

It's true, that the sentences are formulated from the opposite side of the view in the place, that is focused on `@Attribute()`. But I am not sure if their re-formulation from the `@Attribute()` point of view would sound better.

Reviewer 2: Ah! I see what you mean. What if you changed this lasts sentence to recap/rephrase what we're teaching the user? Something like:

Remember, use `@Input()` when you want to keep track of the attribute value and update the associated property. Use `@Attribute()` when you want to pass the value of an HTML attribute to a component or directive constructor.

Reviewer 3: I think that it's very valuable to add also:  ..., it means the ability to have it accessible even before the first change detection happens, what's not available with `@Input()`.

Also, I would use the word `inject` instead of `pass`.

Author: Very valuable discussion 🙏 
I replaced the last section with the 'remember' suggestion.
